openpype/tools/workfiles/lib.py

```diff
@@ -0,0 +1,214 @@
+        self._id = item_id
+        self._directory = directory
+        self._filepath = os.path.join(directory, data["filename"])
+
+    @property
+    def directory(self):
+        return self._directory
+
+    @property
+    def filepath(self):
+        return self._filepath
+
+    @property
+    def id(self):
+        return self._id
+
+    @property
+    def size(self):
+        if os.path.exists(self.filepath):
+            s = os.stat(self.filepath)
+            return s.st_size
+        return 0
+
+
+class TempPublishFiles(object):
+    """Directory where """
+    minute_in_seconds = 60
+    hour_in_seconds = 60 * minute_in_seconds
+    day_in_seconds = 24 * hour_in_seconds
+
+    def __init__(self):
+        root_dir = appdirs.user_data_dir(
+            "published_workfiles_temp", "openpype"
+        )
+        if not os.path.exists(root_dir):
+            os.makedirs(root_dir)
+
+        metadata_path = os.path.join(root_dir, "metadata.json")
+        lock_path = os.path.join(root_dir, "lock.json")
+
+        self._root_dir = root_dir
+        self._metadata_path = metadata_path
+        self._lock_path = lock_path
+        self._log = None
+
+    @property
+    def log(self):
+        if self._log is None:
+            self._log = logging.getLogger(self.__class__.__name__)
+        return self._log
+
+    @property
+    def life_time(self):
+        return int(self.hour_in_seconds)
+
+    @property
+    def size(self):
+        size = 0
+        for item in self.get_items():
+            size += item.size
+        return size
+
+    def add_file(self, src_path):
+        filename = os.path.basename(src_path)
+
+        item_id = str(uuid.uuid4())
+        dst_dirpath = os.path.join(self._root_dir, item_id)
+        if not os.path.exists(dst_dirpath):
+            os.makedirs(dst_dirpath)
+
+        dst_path = os.path.join(dst_dirpath, filename)
+        shutil.copy(src_path, dst_path)
+
+        now = time.time()
+        item_data = {
+            "filename": filename,
+            "expiration": now + self.life_time,
+            "created": now
+        }
+        with self._modify_data() as data:
+            data[item_id] = item_data
+
+        return TempPublishFilesItem(item_id, item_data, dst_dirpath)
+
+    @contextlib.contextmanager
+    def _modify_data(self):
+        start_time = time.time()
+        timeout = 3
+        while os.path.exists(self._lock_path):
+            time.sleep(0.01)
+            if start_time > timeout:
+                self.log.warning((
+                    "Waited for {} seconds to free lock file. Overriding lock."
+                ).format(timeout))
+
+        with open(self._lock_path, "w") as stream:
+            json.dump({"pid": os.getpid()}, stream)
+
+        try:
+            data = self._get_data()
+            yield data
+            with open(self._metadata_path, "w") as stream:
+                json.dump(data, stream)
+
+        finally:
+            os.remove(self._lock_path)
```

Comment:
What's with the lock file. Why do we need to go through this locking mechanism? Would be great to clarify in a docstring or alike. It feels redundant when glancing over the code quickly.

Reply:
Making sure that the metadata file is locked for whole process of adding data. Will fill docs.

Reply:
> Also, what happens if you have a different published workfile with the same name that you are opening (e.g. from another project if e.g. project name isn't in the published filename?)

Each opened publish file will create new item with new id so even if you would open same file twice it would create 2 items and store it to a different place. The metadata just keep track of when they were created and when they should be removed.

> And instead opt for simple yet effective - copy the file once and just open it directly. Leave it up to the user to do whatever they'd like at that moment of time.

We want avoid to give option to open the published workfile directly but want give option to just open it and not save in any workdir (at least from what I understood). Give option to save it is in plan, but in different PR.

Reply:
I understand we'd want to avoid opening the published file for safety reasons.

I'm just worried that the "temp file" logic in place now is worth the amount of code involved to make this work. It means we'll need to maintain the functionality and I wonder whether it provides enough "benefits" over just allowing a user to copy a published workfile back into the workfiles as a new version.

Aside of potentially easier "debugging" of the published workfile I'm personally doubting whether this warrants being this much extra code to maintain. It feels like feature creep. Whereas just saving as another workfile potentially would need only a few extra lines of codes. It could even be a `Loader` in that case! (But I can understand why having it visually in Work Files might be nice).